janus_server/src/client.rs

```diff
@@ -29,63 +37,215 @@ static CLIENT_USER_AGENT: &str = concat!(
+    pub async fn upload(&self, measurement: &V::Measurement) -> Result<(), Error> {
+        let input_shares = self
+            .vdaf_client
+            .shard(&self.vdaf_public_parameter, measurement)?;
 
-    pub async fn aggregator_hpke_sender(
-        http_client: &reqwest::Client,
-        task_id: TaskId,
-        aggregator_endpoint: Url,
-    ) -> Result<HpkeSender, Error> {
-        let hpke_config_response = http_client
-            .get(aggregator_endpoint.join("hpke_config")?)
+        // All supported VDAFs use two aggregators
+        assert_eq!(input_shares.len(), 2);
+
+        let nonce = Nonce {
+            time: Time::from_naive_date_time(self.clock.now()),
+            rand: rand::random(),
+        };
+
+        // No extensions supported yet
+        let extensions = vec![];
+
+        let associated_data = Report::associated_data(nonce, &extensions);
+
+        // Leader's share MUST be the first in the report. We assume/guess that
+        // the first element in input_shares is the leader's. This is true for
+        // all prio3 VDAFs, but not guaranteed generically.
+        // https://github.com/cjpatton/vdaf/issues/40
+        let encrypted_input_shares: Vec<HpkeCiphertext> =
+            [&self.leader_report_sender, &self.helper_report_sender]
+                .into_iter()
+                .zip(input_shares)
+                .map(|(hpke_sender, input_share)| {
+                    Ok(hpke_sender.seal(&input_share.get_encoded(), &associated_data)?)
+                })
+                .collect::<Result<_, Error>>()?;
+
+        let report = Report {
+            task_id: self.task_id,
+            nonce,
+            extensions,
+            encrypted_input_shares,
+        };
+
+        let upload_response = self
+            .http_client
+            .post(self.leader_url.join("upload")?)
+            .body(report.get_encoded())
             .send()
             .await?;
-        let status = hpke_config_response.status();
+        let status = upload_response.status();
         if !status.is_success() {
+            // TODO: decode an RFC 7807 problem document once #25 / #31 land
             return Err(Error::Http(status));
         }
 
-        let hpke_config = HpkeConfig::decode(&mut Cursor::new(
-            hpke_config_response.bytes().await?.as_ref(),
-        ))?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        hpke::HpkeRecipient, time::tests::MockClock, trace::test_util::install_trace_subscriber,
+    };
+    use assert_matches::assert_matches;
+    use mockito::mock;
+    use prio::vdaf::prio3::{Prio3Aes128Count, Prio3Aes128Sum};
 
-        Ok(HpkeSender {
+    fn setup_client<V: VdafClient>(
+        vdaf_client: V,
+        public_parameter: V::PublicParam,
+    ) -> Client<V, V::PublicParam, MockClock> {
+        let task_id = TaskId::random();
+        let clock = MockClock::default();
+        let leader_hpke_recipient =
```

Comment:
nit: it looks like there are a number of parameters shared between the `HpkeRecipient` and the `HpkeSender` that I suppose need to match up. Could we refactor creation of the `HpkeSender` so that it's created from a `TaskId` & `HpkeRecipient`? e.g. `HpkeSender::new(task_id, recipient)`, or perhaps `HpkeSender::to` if that name isn't too generic.

Reply:
I've added `HpkeSender::from_recipient` and `HpkeSender::new`. The former does what you described (though it doesn't take `TaskId` because that can be copied from the `HpkeRecipient`). I didn't implement it as trait `From` because AFAICT we don't yet have any use case for generically transforming an `HpkeRecipient` into an `HpkeSender`, and this way callers don't ever have to worry about disambiguating which `From` implementation they want. My intuition and taste on when to implement traits like `From` is still evolving.

`HpkeSender::new` is a pretty dopey function, but it allowed me to make all the fields on `HpkeSender` private, which I think is defensible. I kinda wish Rust had a `write-only` sort of visibility modifier, so that struct definitions could be used to create trivial constructors but then forbid later mutation or reading of fields.

Reply:
I'm totally OK with not doing a From implementation; I think a From implementation on `HpkeRecipient` would consume the recipient, which isn't what we want in this case. (though I think you could do `From<&HpkeRecipient>`? but that seems pretty unnatural to me, though my own intuition is also not very strong)

+1 to making fields private, even with the additional overhead of needing to write a static constructor method & possibly accessors. If nothing else, it might ease refactorings later, and will make it possible to enforce invariants on the internal structures. I should probably go through and apply this everywhere... we've got a lot of structs that currently make their internal state public without a good reason (other than it's easy).